examples/arkode/C_serial/ark_harmonic_symplectic.c

```diff
@@ -0,0 +1,366 @@
+  sunrealtype Tf          = SUN_RCONST(0.0);
+  sunrealtype dt          = SUN_RCONST(0.0);
+  sunrealtype dTout       = SUN_RCONST(0.0);
+  const sunrealtype T0    = SUN_RCONST(0.0);
+  const sunrealtype A     = SUN_RCONST(10.0);
+  const sunrealtype phi   = SUN_RCONST(0.0);
+  const sunrealtype omega = SUN_RCONST(1.0);
+
+  /* Parse the command line arguments */
+  if (ParseArgs(argc, argv, &args)) { return 1; };
+
+  /* Default integrator options and problem parameters */
+  order            = args.order;
+  use_compsums     = args.use_compsums;
+  num_output_times = args.num_output_times;
+  Tf               = args.Tf;
+  dt               = args.dt;
+  dTout            = (Tf - T0) / ((sunrealtype)num_output_times);
+
+  /* Default problem parameters */
+
+  /* Create the SUNDIALS context object for this simulation */
+  retval = SUNContext_Create(NULL, &sunctx);
+  if (check_retval(&retval, "SUNContext_Create", 1)) return 1;
+
+  printf("\n   Begin simple harmonic oscillator problem\n\n");
+
+  /* Allocate and fill udata structure */
+  udata.A     = A;
+  udata.phi   = phi;
+  udata.omega = omega;
+
+  /* Allocate our state vector [x, v]^T */
+  y        = N_VNew_Serial(2, sunctx);
+  solution = N_VClone(y);
+
+  /* Fill the initial conditions (x0 then v0) */
+  ydata    = N_VGetArrayPointer(y);
+  ydata[0] = A * cos(phi);
+  ydata[1] = -A * omega * sin(phi);
+
+  /* Create SPRKStep integrator */
+  arkode_mem = SPRKStepCreate(xdot, vdot, T0, y, sunctx);
+
+  retval = SPRKStepSetOrder(arkode_mem, order);
+  if (check_retval(&retval, "SPRKStepSetOrder", 1)) { return 1; }
+
+  retval = SPRKStepSetUserData(arkode_mem, &udata);
+  if (check_retval(&retval, "SPRKStepSetUserData", 1)) { return 1; }
+
+  retval = SPRKStepSetUseCompensatedSums(arkode_mem, use_compsums);
+  if (check_retval(&retval, "SPRKStepSetUseCompensatedSums", 1)) { return 1; }
+
+  retval = SPRKStepSetFixedStep(arkode_mem, dt);
+  if (check_retval(&retval, "SPRKStepSetFixedStep", 1)) { return 1; }
+
+  retval = SPRKStepSetMaxNumSteps(arkode_mem, ((long int)ceil(Tf / dt)) + 2);
+  if (check_retval(&retval, "SPRKStepSetMaxNumSteps", 1)) { return 1; }
+
+  /* Print out starting energy, momentum before integrating */
+  tret = T0;
+  tout = T0 + dTout;
+  fprintf(stdout, "t = %.6Lf, x(t) = %.6Lf, E = %.6Lf, sol. err = %.6Lf\n",
+          (long double)tret, (long double)ydata[0],
+          (long double)Energy(y, dt, &udata), (long double)SUN_RCONST(0.0));
+
+  /* Do integration */
+  for (iout = 0; iout < num_output_times; iout++)
+  {
+    if (args.use_tstop) { SPRKStepSetStopTime(arkode_mem, tout); }
+    retval = SPRKStepEvolve(arkode_mem, tout, y, &tret, ARK_NORMAL);
+
+    /* Compute the anaytical solution */
+    Solution(tret, y, solution, &udata);
+
+    /* Compute L2 error */
+    N_VLinearSum(SUN_RCONST(1.0), y, -SUN_RCONST(1.0), solution, solution);
+    err = sqrt(N_VDotProd(solution, solution));
+
+    /* Output current integration status */
```

Comment:
Since we output the Energy and it should remain constant, this seems redundant to print out.

Reply:
There will be some error though since it's not conserved exactly and it's easier to see this in the error (or if we printed more digits of the energy)

Reply:
I like printing more digits of the energy instead because the error is not as clearly defined.

Reply:
The error would be the difference from the initial energy

Reply:
Well, yes, of course.  But my point is that you can observe conservation and the actual value for the energy just by looking at energy (i.e. error is not necessary, just more digits).

Reply:
Yes, they're both ways of showing the same thing. I think it's easier to see an error that's something like `1e-15`  but either way is good.